File: src/main/java/gargoyle/rpycg/ui/model/DROPPING.java
package gargoyle.rpycg.ui.model;

import javafx.scene.control.TreeCell;
import javafx.scene.input.DragEvent;

public enum DROPPING {
    ONTO, ABOVE, BELOW;

    @SuppressWarnings("ConstantExpression")
    public static DROPPING determineDropping(final TreeCell<DisplayItem> cell, final DragEvent event,
                                             final double bond) {
        final double ratio = event.getY() / cell.getLayoutBounds().getHeight();
        if (ratio < bond) {
            return ABOVE;
        }
        if (ratio > 1 - bond) {
            return BELOW;
        }
        return ONTO;
    }
}

File: app/src/main/kotlin/gargoyle/rpycg/ui/model/DROPPING.kt
package gargoyle.rpycg.ui.model

import javafx.scene.control.TreeCell
import javafx.scene.input.DragEvent

enum class DROPPING {
    ONTO, ABOVE, BELOW;

    companion object {
        fun determineDropping(cell: TreeCell<DisplayItem>, event: DragEvent, bond: Double): DROPPING {
            val ratio = event.y / cell.layoutBounds.height
            return if (ratio < bond) {
                ABOVE
            } else if (ratio > 1 - bond) {
                BELOW
            } else ONTO
        }
    }
}
